### PR TITLE
[API] Add Project endpoint "Dicoms"

### LINKED ISSUE #6774: [API] New Project /dicoms endpoint
**Is your feature request related to an existing module?**
This feature is related to the API

**Describe the solution you'd like**
It would be useful to add the `endpoint /projects/<project>/dicoms` that would give access to a complete list of all DICOM tarfiles contained in all projects. It is similar to all other `Project` endpoints, for example `<host>/api/v0.0.3/projects/rye/images`

### PR DESCRIPTION
**Pull request by spell00**
## Brief summary of changes
This Pull Request adds an endpoint at /projects/\<project\>/dicoms. It returns a complete list of DICOM files contained in a project (and other pertinent informations)

#### Testing instructions
Go to https://\<hostname\>/api/v0.0.3/projects/rye/dicoms in a browser. The page should return a JSON in a string with information on all DICOM contained in the project

#### Link(s) to related issue(s)

* Resolves #6774

### CODE REVIEW

**Comment by christinerogers** [comment]
@spell00 Is this PR up to date?  
If so, @xlecours do you have time to review this? if not let me know and I'll find someone perhaps Cecile.

File: modules/api/php/models/projectimagesrow.class.inc [hunk 1]
```diff
@@ -1,4 +1,4 @@
-<?php declare(strict_types=1);
+<?php
```
Comment (xlecours) [inline modules/api/php/models/projectimagesrow.class.inc:1]: Why is this line changed?

File: modules/api/php/provisioners/projectdicomsrowprovisioner.class.inc [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+             LEFT JOIN tarchive_series ts
+               ON (ts.TarchiveID = t.TarchiveID)
+	     LEFT JOIN server_processes sp
+               ON (sp.pid = f.ProcessProtocolID)
+             WHERE
+               c.Active = \'Y\' AND
+               s.Active = \'Y\' AND
+               project.Name = :v_projectname AND
+               f.InsertTime > :v_time
+             ORDER BY f.InsertTime ASC
```
Comment (cmadjar) [inline modules/api/php/provisioners/projectdicomsrowprovisioner.class.inc:81]: I would simplify the query for this endpoint. I guess the goal of the endpoint is to have a DICOMs project endpoint that lists all the tarchive files in the database. In that case, there is no need to join to the files table or the tarchive_series table. Below is the simplification I was thinking about.

```suggestion
             SELECT
               s.CandID as Candidate,
               c.PSCID as PSCID,
               c.Entity_type as Entity_type,
               s.Visit_label as Visit,
               s.Date_visit as Visit_date,
               s.CenterID as CenterID,
               p.Name as Site,
               t.DateAcquired as date_acquired,
               t.DateFirstArchived as date_first_archived,
               t.DateLastArchived as date_last_archived,
               t.TarchiveID as tarchiveid,
               t.DicomArchiveID as DicomArchiveID,
               t.ArchiveLocation as Archive,
	           t.SourceLocation as Source,
             FROM
               tarchive t
             LEFT JOIN  session s
               ON (t.SessionID = s.ID)
             LEFT JOIN candidate c
               ON (s.CandID = c.CandID)
             LEFT JOIN psc p
               ON (s.CenterID = p.CenterID)
             LEFT JOIN Project project
               ON (c.RegistrationProjectID = project.ProjectID)
             WHERE
               c.Active = \'Y\' AND
               s.Active = \'Y\' AND
               project.Name = :v_projectname AND
               t.DateLastArchived > :v_time
             ORDER BY t.DateLastArchived ASC
```

**Review by cmadjar — DISMISSED** [review]
@spell00 Nice work! 

See my comment on the query. Let me know if I missed something or misunderstood the goal of the endpoint. Thanks!

File: modules/api/php/provisioners/projectdicomsrowprovisioner.class.inc (hunk 2 above)
Comment (spell00) [reply modules/api/php/provisioners/projectdicomsrowprovisioner.class.inc:81]: yes, that is the idea. I'll have a look at this, thank you for the suggestion!

File: modules/api/php/models/projectdicomsrow.class.inc [hunk 3]
```diff
@@ -0,0 +1,123 @@
+<?php
```
[the remark this answers is not in the document]
Comment (spell00) [reply modules/api/php/models/projectdicomsrow.class.inc:1]: I was told it was not necessary to declare strict_types like this in Loris. Even if this is the case, I'm putting it back for consistency within that module (it is declared in other php files)

File: modules/api/php/provisioners/projectdicomsrowprovisioner.class.inc [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+               t.DateFirstArchived as date_first_archived,
+               t.DateLastArchived as date_last_archived,
+               t.TarchiveID as tarchiveid,
+               t.DicomArchiveID as DicomArchiveID,
+               t.ArchiveLocation as Archive,
+	       t.SourceLocation as Source,
+             FROM
+               tarchive t
+             LEFT JOIN  session s
+               ON (t.SessionID = s.ID)
```
Comment (spell00) [inline modules/api/php/provisioners/projectdicomsrowprovisioner.class.inc:65]: there are NULL values in the table `tarchive` in raisinbread, is this normal?

File: modules/api/php/provisioners/projectdicomsrowprovisioner.class.inc [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+ *
+ * PHP Version 7
+ *
+ * @category API
+ * @package  Loris
+ * @author   Xavier Lecours Boucher <xavier.lecours@mcin.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+class ProjectDicomsRowProvisioner extends DBObjectProvisioner
```
Comment (spell00) [inline modules/api/php/provisioners/projectdicomsrowprovisioner.class.inc:31]:
```suggestion
class ProjectDicomsObjectProvisioner extends DBObjectProvisioner
```

File: modules/api/php/provisioners/projectdicomsrowprovisioner.class.inc [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+     * Create a RowProvisioner
+     *
+     * @param \Project  $project   The project from which dicoms are requested
+     * @param \DateTime $since     The date from which dicoms are requested
+     * @param string    $classname The class name of the returned objects
+     */
+    function __construct(
+        \Project $project,
+        \DateTime $since,
+        string $classname='default'
```
Comment (spell00) [inline modules/api/php/provisioners/projectdicomsrowprovisioner.class.inc:43]:
```suggestion
        string $classname='ProjectDicomsObjectProvisioner'
```

Comment (xlecours) [reply modules/api/php/provisioners/projectdicomsrowprovisioner.class.inc:43]: That would be `\LORIS\api\Models\ProjectDicomsRow.class.inc`

File: modules/api/php/provisioners/projectdicomsrowprovisioner.class.inc [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+    /**
+     * Returns an instance of a ProjectDicomsRow object for a given
+     * table row.
+     *
+     * @param array $row The database row from the LORIS Database class.
+     *
+     * @return \LORIS\Data\DataInstance An instance representing this row.
+     */
+    public function getInstance($row) : \LORIS\Data\DataInstance
```
Comment (xlecours) [inline modules/api/php/provisioners/projectdicomsrowprovisioner.class.inc:95]: DBObjectPRovisioners do not need a getInstance function.

File: modules/api/php/models/projectdicomsrow.class.inc [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    private $_processid;
+    private $_acquisitionprotocolid;
+    private $_tarchiveseriesid;
+    private $_processprotocolid;
+    /**
+     * Create a new ProjectDicomsRow
+     *
+     * @param array $row An array of image propeties
+     */
+    public function __construct(array $row)
```
Comment (xlecours) [inline modules/api/php/models/projectdicomsrow.class.inc:47]: That constructor is never called.
PDO:statement fetchall instanciate the objects and set their values using `__set`.

File: modules/api/php/endpoints/project/dicoms.class.inc [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        } catch (\Exception $e) {
+            return new \LORIS\Http\Response\JSON\BadRequest(
+                $e->getMessage()
+            );
+        }
+
+        $provisioner = new \LORIS\api\Provisioners\ProjectDicomsRowProvisioner(
+            $this->_project,
+            $since
+        );
```
Comment (xlecours) [inline modules/api/php/endpoints/project/dicoms.class.inc:125]:
```suggestion
        $provisioner = new \LORIS\api\Provisioners\ProjectDicomsObjectProvisioner(
            $this->_project,
            $since,
            '\LORIS\api\Models\ProjectDicomsRow'
        );
```

File: modules/api/php/provisioners/projectdicomsrowprovisioner.class.inc (hunk 5 above)
Comment (xlecours) [reply modules/api/php/provisioners/projectdicomsrowprovisioner.class.inc:31]: YOu will need to change the filename as well.

File: modules/api/php/provisioners/projectdicomsrowprovisioner.class.inc (hunk 4 above)
Comment (christinerogers) [reply modules/api/php/provisioners/projectdicomsrowprovisioner.class.inc:65]: which columns?  
@spell00 can you pinpoint the rows you think might be problematic and we'll take a look? 
please link to the sql lines in raisinbread on aces/loris for common reference.

Comment (cmadjar) [reply modules/api/php/provisioners/projectdicomsrowprovisioner.class.inc:65]: Yes, my guess is these are the few that I had to add in the raisinbread where the PSCID and CandID in the PatientName did not refer to an existing candidate (hence, no MINC images inserted into the files table).

Basically, because of the bad PatientName present in the DICOMs, the pipeline cannot associate them to a given candidate/session and therefore their SessionID is NULL. I am not worried about those cases as typically, projects would end up removing those entries, relabelling the DICOMs correctly and reupload them with the correct PatientName.

Hope that makes sense. @spell00 let me know if that is not clear.

**Comment by spell00** [comment]
now I get `{"Dicoms":{}}` when trying to access `<host>/api/v0.0.3/projects/rye/dicoms`

**Comment by christinerogers** [comment]
@xlecours and @kongtiaowang  - would you have a few minutes to (re)review this PR this week?

File: modules/api/php/endpoints/project/dicoms.class.inc [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+            );
+        }
+
+        $provisioner = new \LORIS\api\Provisioners\ProjectDicomsObjectProvisioner(
+            $this->_project,
+            $since,
+            '\LORIS\api\Models\ProjectDicomsObject'
+        );
+
+        $all = $provisioner->getAllInstances();
```
Comment (xlecours) [inline modules/api/php/endpoints/project/dicoms.class.inc:128]: `$all` ?? all what?

File: modules/api/php/endpoints/project/dicoms.class.inc [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+    /**
+     * Create an array representation of this endpoint's reponse body
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    private function _handleGET(ServerRequestInterface $request): ResponseInterface
+    {
+        if (isset($this->_cache)) {
```
Comment (xlecours) [inline modules/api/php/endpoints/project/dicoms.class.inc:109]: Make sure that _hasAccess is called
The dicom_archives module requires `dicom_archive_view_allsites` to access it. 
@cmadjar The same applies here right?

File: modules/api/php/endpoints/project/dicoms.class.inc (hunk 10 above)
Comment (spell00) [reply modules/api/php/endpoints/project/dicoms.class.inc:128]: `$provisioner->getAllInstances` , so obviously all instances (of dicoms object). I get the point you are making though, I'll give it a more meaningful name, like `$allDicomsInstances`

File: modules/api/php/endpoints/project/dicoms.class.inc (hunk 11 above)
Comment (spell00) [reply modules/api/php/endpoints/project/dicoms.class.inc:136]: I added it, but I did like for the `project/recordings` [endpoint](https://github.com/aces/Loris/blob/main/modules/api/php/endpoints/project/recordings.class.inc#L97), which calls `_hasAccess` in `handle`. It would not matter here since I don't think the endpoints in `\projects` will ever have any other handles than handleGET, but I think it's better in `handle` for consistency

File: modules/api/php/endpoints/project/dicoms.class.inc (hunk 10 above)
Comment (spell00) [reply modules/api/php/endpoints/project/dicoms.class.inc]: now that I'm using `iterator_to_array` as you suggested, I don't think this variable is necessary anymore, I did `$dicoms = iterator_to_array($provisioner->getAllInstances());`, which is consistent with the others endpoints (e.g. for [images](https://github.com/aces/Loris/blob/a87ead08684d8f4be4e10a982339eae440b075e7/modules/api/php/endpoints/project/images.class.inc#L131), the final array containing all info to display about images in a project is called `$images` )

**Comment by christinerogers** [comment]
@spell00 this seems like a minor change request -- can you address it and move this forward today?

**Comment by spell00** [comment]
ok @christinerogers , I'll take care of it today

File: modules/api/php/endpoints/candidate/visit/electrophysiology/channels.class.inc [hunk 12]
```diff
@@ -143,4 +143,4 @@ class Channels extends Endpoint implements \LORIS\Middleware\ETagCalculator
     {
         return md5(json_encode($this->_handleGET($request)->getBody()));
     }
-}
```
Comment (spell00) [inline modules/api/php/endpoints/candidate/visit/electrophysiology/channels.class.inc:146]: I tried to remove this change but I can't. They seem to be exactly the same, I did not add a newline... same for the other similar changes in this PR

**Comment by christinerogers** [comment]
@spell00 is this ready for re-review? it's not clear if you're still working on resolving the comment you added.

**Comment by spell00** [comment]
@christinerogers It is ready for review. 

Earlier, I was trying to point out that in the changed files, there is indications that I removed the last line in some files. I did not remove the last line and I can't get rid of the indication there is a change. I was trying to get things faster, I am expecting having a comment about these.

File: modules/api/php/models/candidatesrow.class.inc [hunk 13]
```diff
@@ -24,7 +24,6 @@ class CandidatesRow implements \LORIS\Data\DataInstance
 {
     private $_candid;
     private $_projectname;
-    private $_projectid;
```
Comment (xlecours) [inline modules/api/php/models/candidatesrow.class.inc:27]: Why is this removed? It is used in https://github.com/aces/Loris/pull/6775/files#diff-e8883d6dbf607c4d208c6405a42a0f5b9d044d9e2e705ec5cfaca41703943627R91

**Review by xlecours — CHANGES_REQUESTED** [review]

File: modules/api/php/provisioners/projectdicomsobjectprovisioner.class.inc [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+     * Create a ObjectProvisioner
+     *
+     * @param \Project  $project   The project from which dicoms are requested
+     * @param \DateTime $since     The date from which dicoms are requested
+     * @param string    $classname The class name of the returned objects
+     */
+    function __construct(
+        \Project $project,
+        \DateTime $since,
+        string $classname='ProjectDicomsObjectProvisioner'
```
Comment (xlecours) [inline modules/api/php/provisioners/projectdicomsobjectprovisioner.class.inc:42]: I think this should be `ProjectDicomsObject`
And ProjectDicoms would be a good class name since they are object.

**Review by xlecours — CHANGES_REQUESTED** [review]

**Comment by spell00** [comment]
blocked by #7101

**Comment by spell00** [comment]
@xlecours @driusan It is now passing Travis and ready for final review

File: modules/api/docs/LorisRESTAPI.md [hunk 15]
```diff
@@ -263,6 +263,36 @@ will return a JSON object of the form:
 }
 ```
 
+#### 2.1.6 Single project dicoms tar files
```
Comment (xlecours) [inline modules/api/docs/LorisRESTAPI.md:266]: I don't think this file can be changed since the endpoint is available fro 0.0.4-dev only.

File: modules/api/php/provisioners/projectdicomsobjectprovisioner.class.inc [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+     * Create a ObjectProvisioner
+     *
+     * @param \Project  $project   The project from which dicoms are requested
+     * @param \DateTime $since     The date from which dicoms are requested
+     * @param string    $classname The class name of the returned objects
+     */
+    function __construct(
+        \Project $project,
+        \DateTime $since,
+        string $classname='ProjectDicomsObject'
```
Comment (xlecours) [inline modules/api/php/provisioners/projectdicomsobjectprovisioner.class.inc:42]: classname is `LORIS\api\Models\ProjectDicoms`

**Review by xlecours — CHANGES_REQUESTED** [review]

File: modules/api/php/provisioners/projectdicomsobjectprovisioner.class.inc (hunk 16 above)
Comment (spell00) [reply modules/api/php/provisioners/projectdicomsobjectprovisioner.class.inc]: I updated it to:  `string $classname='ProjectDicoms'`, but did you want it to be  `string $classname='LORIS\api\Models\ProjectDicoms'` ?

**Review by xlecours — APPROVED** [review]
I wouldn't mind having @cmadjar backing on this. Are the properties of the Dicoms items ok?

File: modules/api/docs/LorisRESTAPI_v0.0.4-dev.md [hunk 17]
```diff
@@ -263,6 +263,37 @@ will return a JSON object of the form:
...
+      "tarchiveid": "27",
+      "DicomArchiveID": "1.2.840.113745.101000.1022000.39911.6153.5242769",
+      "Archive": "2018/DCM_2018-04-20_ImagingUpload-13-42-zcoZR0.tar",
+      "Source": "/tmp/ImagingUpload-13-42-zcoZR0",
+      "FileName": "DCM_2018-04-20_ImagingUpload-13-42-zcoZR0.tar"
+    },
+    ...
+  ]
+}
+```
```
Comment (xlecours) [inline modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:295]: There should be a mention about the since parameter.

**Review by xlecours — DISMISSED** [review]

File: modules/api/docs/LorisRESTAPI_v0.0.4-dev.md (hunk 17 above)
Comment (spell00) [reply modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:295]: what is the since parameter?

Comment (xlecours) [reply modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:295]: https://github.com/aces/Loris/pull/6775/files#diff-abcc8bdf59aa52372d7e9df59277bafe7a4479d355250919912ff4a087939113R135

Comment (driusan) [reply modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:305]: I'm confused by this thread.. it seems to be mentioned?

Comment (spell00) [reply modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:305]: I added something about the since parameter after xlecours comment

Comment (christinerogers) [reply modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:305]: thanks @spell00  - @xlecours could you take a look when you have a moment? this feels close to success :)

**Comment by cmadjar** [comment]
@xlecours can you take this PR over please? Thank you :)

**Comment by ridz1208** [comment]
@cmadjar it seems like @xlecours 's next to latest review approved the changes and asked you to take a second look at it. I think this one is also in your court in that case

from @xlecours 
> I wouldn't mind having @cmadjar backing on this. Are the properties of the Dicoms items ok?

**Comment by ridz1208** [comment]
@xlecours @cmadjar just rebased the PR

File: modules/api/docs/LorisRESTAPI_v0.0.4-dev.md [hunk 18]
```diff
@@ -263,6 +263,47 @@ will return a JSON object of the form:
...
+      "CenterID": "1",
+      "Site": "Data Coordinating Center",
+      "date_acquired": "2018-04-20",
+      "date_first_archived": "2019-05-23 13:46:39",
+      "date_last_archived": "2019-05-23 13:46:39",
+      "tarchiveid": "27",
+      "DicomArchiveID": "1.2.840.113745.101000.1022000.39911.6153.5242769",
+      "Archive": "2018/DCM_2018-04-20_ImagingUpload-13-42-zcoZR0.tar",
+      "Source": "/tmp/ImagingUpload-13-42-zcoZR0",
+      "FileName": "DCM_2018-04-20_ImagingUpload-13-42-zcoZR0.tar"
```
Comment (cmadjar) [inline modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:290]: is there any reason why the following fields are not returned?
```
AcquisitionCount
DicomFileCount
ScannerManufacturer
ScannerModel
ScannerSoftware
ScannerSerialNumber
```

Those are fields I could see as being informative to the user accessing the DICOM archives.

File: modules/api/php/provisioners/projectdicomsobjectprovisioner.class.inc [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+               s.Date_visit as Visit_date,
+               s.CenterID as CenterID,
+               p.Name as Site,
+               t.DateAcquired as date_acquired,
+               t.DateFirstArchived as date_first_archived,
+               t.DateLastArchived as date_last_archived,
+               t.TarchiveID as tarchiveid,
+               t.DicomArchiveID as DicomArchiveID,
+               t.ArchiveLocation as Archive,
+	           t.SourceLocation as Source
```
Comment (cmadjar) [inline modules/api/php/provisioners/projectdicomsobjectprovisioner.class.inc:60]:
```suggestion
	             t.SourceLocation as Source
```

File: modules/api/docs/LorisRESTAPI_v0.0.4-dev.md (hunk 18 above)
Comment (cmadjar) [reply modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:290]: Also, I notice that in the JSON response, keys have all kinds of different convention. I would tend to make them consistent (either stick to CamelCase, or _ separated values, consistent casing etc...). 

Sorry, my brain gets easily distracted with inconsistencies ;).

**Review by cmadjar — CHANGES_REQUESTED** [review]
@ridz1208 thank you for pinging me on this, I did not see the previous comment. See attached minor suggestions. Let me know your thoughts.

**Comment by ridz1208** [comment]
@xlecours this one is back in your court

**Comment by cmadjar** [comment]
@xlecours this one is in your court. Should be straightforward though?

**Comment by xlecours** [comment]
This still needs work. 
1. The property names in the response should be case-consistent
2. I think we should hide : CenterID, tarchiveid and  Source since they are only meaningful from within the database of on the filesystem.
3. There should be a link to the resource (where it can be downloaded) 
3. Documentation should be add in the api specification of the api module (schema.yml)

On a side note: could this be done using /dicom_archive instead?

**Comment by ridz1208** [comment]
@cmadjar assigning to you to bring up at Imaging meeting see if this is still needed

**Comment by cmadjar** [comment]
Discussed during the imaging call of Jul 4th. It is stale and needs work that will not be done in the short term.
